File: 2019/day02-program-alarm/intcode.py
#!/usr/bin/env python

def readInput() -> list:
    with open('day02.in') as f:
        return list(map(int, f.read().strip().split(',')))


def simulate(inA: int = 12, inB: int = 2) -> int:
    program = readInput()
    program[1], program[2] = inA, inB
    run(program)
    return program[0]


def run(program: list) -> list:
    for i in range(0, len(program), 4):
        if program[i] == 99:
            break
        elif program[i] == 1:
            program[program[i+3]] = program[program[i+1]] + \
                program[program[i+2]]
        elif program[i] == 2:
            program[program[i+3]] = program[program[i+1]] * \
                program[program[i+2]]
    return program


def nounVerb() -> int:
    program = readInput()
    noun, verb = findInputs(program, target=19690720)
    return 100 * noun + verb


def findInputs(program: list, target: int) -> (int, int):
    for a in range(100):
        lo, hi = 0, 99
        while lo <= hi:
            b = lo + (hi - lo) // 2
            res = simulate(a, b)
            if res == target:
                return (a, b)
            elif res < target:
                lo = b + 1
            else:
                hi = b - 1
    return (0, 0)


if __name__ == '__main__':
    print(f"Part 1 = {simulate()}")
    print(f"Part 2 = {nounVerb()}")

File: 2019/day02-program-alarm/intcode.ts
#!/usr/bin/env ts-node
import { splitAllLinesAsNumberBy } from "@utils";
import * as _ from "lodash";

enum Code {
    ADD = 1,
    MULTIPLY = 2,
    HALT = 99,
}

const getFirstPosition = (program: number[], noun: number, verb: number): number => {
    restoreGravityAssist(program, noun, verb);
    run(program);
    return program[0];
};

const restoreGravityAssist = (program: number[], noun: number, verb: number) => {
    program[1] = noun;
    program[2] = verb;
};

const run = (program: number[]) => {
    const op = (code: Code) => (code === Code.ADD ? _.add : _.multiply);
    for (let i = 0; ; i += 4) {
        let [code, in1, in2, out] = program.slice(i, i + 4);
        if (code === Code.HALT) break;
        program[out] = op(code)(program[in1], program[in2]);
    }
};

const getNounVerb = (program: number[], target: number): number => {
    const [noun, verb] = search(program, target);
    return 100 * noun + verb;
};

const search = (program: number[], target: number, size = 99): [number, number] => {
    const mid = (min: number, max: number): number => min + Math.floor((max - min) / 2);
    let [output, noun, verb] = [0, 0, 0];
    let [minNoun, maxNoun] = [0, size];

    while (minNoun <= maxNoun) {
        noun = mid(minNoun, maxNoun);
        let [minVerb, maxVerb] = [0, size];

        while (minVerb <= maxVerb) {
            verb = mid(minVerb, maxVerb);
            output = getFirstPosition(_.cloneDeep(program), noun, verb);

            if (output === target) return [noun, verb];
            else if (output < target) minVerb = verb + 1;
            else maxVerb = verb - 1;
        }

        if (output < target) minNoun = noun + 1;
        else maxNoun = noun - 1;
    }
    return [-1, -1];
};

if (require.main === module) {
    const program: number[] = splitAllLinesAsNumberBy("day02.in", ",")[0];
    const [a, b] = ["Part 1", "Part 2"];
    console.time(a);
    console.timeLog(a, getFirstPosition(_.cloneDeep(program), 12, 2));
    console.time(b);
    console.timeLog(b, getNounVerb(program, 19690720));
}
